### PR TITLE
pkg/srv, embed, etcdmain: Support multiple clusters in the same DNS domain

### PR DESCRIPTION
**Pull request by tavish-stripe**
This PR adds the ability to configure the subdomain used for DNS cluster discovery. Previously, the domain was hardcoded to `_etcd-server._tcp.FOO` and `_etcd-server-ssl._tcp.FOO`. This did not work with our DNS setup, so I've made this patch to allow configuring the subdomain portion of the SRV queries.

We've been running this in production for a few months.

### CODE REVIEW

**Comment by hexfusion** [comment]
@xiang90, :+1:

**Comment by tavish-stripe** [comment]
@xiang90 / @hexfusion Thanks for taking a look. Let me know how I can help. If this overall approach looks good I'm happy to rebase on master to resolve the conflict.

File: etcdmain/config.go [hunk 1]
```diff
@@ -152,6 +152,8 @@ func newConfig() *config {
 
 	fs.StringVar(&cfg.Dproxy, "discovery-proxy", cfg.Dproxy, "HTTP proxy to use for traffic to discovery service.")
 	fs.StringVar(&cfg.DNSCluster, "discovery-srv", cfg.DNSCluster, "DNS domain used to bootstrap initial cluster.")
+	fs.StringVar(&cfg.DNSClusterServiceNameSSL, "discovery-srv-name-ssl", cfg.DNSClusterServiceNameSSL, "Service name to query when using DNS discovery for SSL")
```
Comment (xiang90) [inline etcdmain/config.go]: Can we reduce the configurable flag down to one? The goal is to be able to have N etcd discovery under one domain, not to make the ssl suffix configurable?

**Comment by xiang90** [comment]
@tavish-stripe I am OK with the motivation.

**Comment by hexfusion** [comment]
@tavish-stripe thank you for your contribution! Could you do me a favor and create an issue for this PR and briefly outline the limitations you faced with your DNS setup and how this PR resolved it? You could copy and paste a lot of what you have above. Would be good to track the issue while we work through this. I hope to have sometime to review after the holiday weekend. Thanks again!

**Comment by hexfusion** [comment]
> Previously, the domain was hardcoded to _etcd-server._tcp.FOO and _etcd-server-ssl._tcp.FOO. This did not work with our DNS setup

@tavish-stripe, just checking in. I am trying to better understand the issues you faced with your DNS setup. For example, why does this type of setup not work for you? It would help me a lot to get some context to why this change is necessary.

### cluster 1
-–discovery-srv=foo.local

;SVR Records
_etcd-server-ssl._tcp.sub.foo.local. 300 IN  SRV  0 0 2380 node1.foo.local.
_etcd-server-ssl._tcp.sub.foo.local. 300 IN  SRV  0 0 2380 node2.foo.local.
_etcd-server-ssl._tcp.sub.foo.local. 300 IN  SRV  0 0 2380 node3.foo.local.

_etcd-client-ssl._tcp.sub.foo.local. 300 IN  SRV  0 0 2379 node1.foo.local.
_etcd-client-ssl._tcp.sub.foo.local. 300 IN  SRV  0 0 2379 node2.foo.local.
_etcd-client-ssl._tcp.sub.foo.local. 300 IN  SRV  0 0 2379 node3.foo.local.

### cluster 2
-–discovery-srv=sub.foo.local

;SVR Records
_etcd-server-ssl._tcp.sub.foo.local. 300 IN  SRV  0 0 2380 node4.sub.foo.local.
_etcd-server-ssl._tcp.sub.foo.local. 300 IN  SRV  0 0 2380 node5.sub.foo.local.
_etcd-server-ssl._tcp.sub.foo.local. 300 IN  SRV  0 0 2380 node6.sub.foo.local.

_etcd-client-ssl._tcp.sub.foo.local. 300 IN  SRV  0 0 2379 node4.sub.foo.local.
_etcd-client-ssl._tcp.sub.foo.local. 300 IN  SRV  0 0 2379 node5.sub.foo.local.
_etcd-client-ssl._tcp.sub.foo.local. 300 IN  SRV  0 0 2379 node6.sub.foo.local.

**Comment by hexfusion** [comment]
ping @tavish-stripe :)

**Comment by tavish-stripe** [comment]
@hexfusion thanks for the ping!

With our internal service configuration setup everything is under the same `foo.local` domain and we can't tack on an extra subdomain on a per-cluster basis, so being able to configure the subdomain purpose is useful for us.

**Comment by hexfusion** [comment]
@tavish-stripe thank you for the reply, so if I understand you correctly your saying that your DNS setup somehow is not configurable to allow sub domains for SRV records. So that all SRV records are defined as the root domain?

So as a work around you are passing the SRV name to etcd to circumvent this? I think that is a crafty solution :) but at that point we are hacking the DNS results which are supposed to be defining the discovery. So this would be a hybrid DNS discovery mechanism one that uses parts of the SRV records and manually overriding others. Please feel free to chime in if my assessment is wrong.

Before we go down this route I would like to propose an alternative discovery mechanism based on TXT records and see if it would be useful for others. I believe it could also solve your problem? See https://github.com/coreos/etcd/issues/9128

/cc @gyuho @xiang90

**Comment by tavish-stripe** [comment]
@hexfusion sorry for the delay.

Since we're using Consul for service discovery internally, it's very easy for us to register each etcd cluster in our consul cluster as a separate service. This allows us to use Consul's automatically generated SRV records for each etcd service in the cluster. As a result, we get SRV services registered at `_etcd-cluster-1._tcp.service.consul`, `_etcd-cluster-2._tcp.service.consul`, etc. Consul does not, as far as I know, allow you to tack on subdomains per service.

Using SRV discovery (with our own provided service names) allows us to easily use Consul DNS to bootstrap our etcd clusters without having to configure DNS specially for each cluster. The TXT solution you suggested does not allow us to leverage Consul's built-in functionality.

> but at that point we are hacking the DNS results which are supposed to be defining the discovery. So this would be a hybrid DNS discovery mechanism one that uses parts of the SRV records and manually overriding others.

It might be helpful to clarify our use case. The solution I'm proposing uses all of the SRV responses and does not override any of them. Here's an example query and response generated by consul. There are two clusters here advertising a client service each (`etcd-client-a` and `etcd-client-b`).

```
$ dig srv _etcd-client-a._tcp.service.consul

;; ANSWER SECTION:
_etcd-client-a._tcp.service.consul. 5 IN  SRV 1 1 2379 etcda-00000000000000000.node.qa-northwest.consul.
_etcd-client-a._tcp.service.consul. 5 IN  SRV 1 1 2379 etcda-11111111111111111.node.qa-northwest.consul.
_etcd-client-a._tcp.service.consul. 5 IN  SRV 1 1 2379 etcda-22222222222222222.node.qa-northwest.consul.

;; ADDITIONAL SECTION:
etcda-00000000000000000.node.qa-northwest.consul. 5 IN A 10.0.0.1
etcda-11111111111111111.node.qa-northwest.consul. 5 IN A 10.0.0.2
etcda-22222222222222222.node.qa-northwest.consul. 5 IN A 10.0.0.3


$ dig srv _etcd-client-b._tcp.service.consul

;; ANSWER SECTION:
_etcd-client-b._tcp.service.consul. 5 IN  SRV 1 1 2379 etcdb-00000000000000000.node.qa-northwest.consul.
_etcd-client-b._tcp.service.consul. 5 IN  SRV 1 1 2379 etcdb-11111111111111111.node.qa-northwest.consul.
_etcd-client-b._tcp.service.consul. 5 IN  SRV 1 1 2379 etcdb-22222222222222222.node.qa-northwest.consul.

;; ADDITIONAL SECTION:
etcdb-00000000000000000.node.qa-northwest.consul. 5 IN A 10.1.0.1
etcdb-11111111111111111.node.qa-northwest.consul. 5 IN A 10.1.0.2
etcdb-22222222222222222.node.qa-northwest.consul. 5 IN A 10.1.0.3
```

Does that help clarify the use case?

**Comment by hexfusion** [comment]
@tavish-stripe this is very helpful will review shortly. Thanks!

**Comment by hexfusion** [comment]
> Since we're using Consul for service discovery internally, it's very easy for us to register each etcd cluster in our consul cluster as a separate service. This allows us to use Consul's automatically generated SRV records for each etcd service in the cluster. As a result, we get SRV services registered at _etcd-cluster-1._tcp.service.consul, _etcd-cluster-2._tcp.service.consul, etc. Consul does not, as far as I know, allow you to tack on subdomains per service.

@tavish-stripe 

- First lets start with consol, it appears that support for multiple domains is a long standing issue. https://github.com/hashicorp/consul/issues/290. Currently the recommended solution by Armon is to have a separate consul per domain. But as this is a one to many relationship of consul/datacenter to etcd cluster, your usage is really the only option for now.

- Second it does seems that using your proposed `instance` naming is supported by https://tools.ietf.org/html/rfc6763 which is good.

The problem from my experience is is that SRV discovery causes a lot of confusion for many folks. So to @xiang90's point we want to keep this to a single flag for simplicity. We would also need to add documentation and tests.

`--discovery-srv-name`

will comment on the code now a bit.

File: etcdmain/config.go (hunk 1 above)
Comment (hexfusion) [reply etcdmain/config.go]: Yes we already know if etcd is using TLS at this point right, how can we get this information from etcd? Can you name your etcd service `_etcd-server-ssl` then in the case of cluster 'a' cumulative SRV would be `_etcd-server-ssl-a._tcp.service.consul. using `--discovery-srv-name a?

File: embed/config.go [hunk 2]
```diff
@@ -391,6 +394,35 @@ func (cfg *Config) PeerURLsMapAndToken(which string) (urlsmap types.URLsMap, tok
 	return urlsmap, token, err
 }
 
+// Run through the various DNS discovery options and defaults.
+func (cfg *Config) GetDNSClusterNames() ([]string, error) {
+	var clusterStrs []string
+	var cerr error
+	var serviceName string
+	var scheme string
```
Comment (hexfusion) [inline embed/config.go]:
```
var (
    clusterStrs []string
    cerr error
    serviceName string
    scheme string
)?
```

**Comment by hexfusion** [comment]
@tavish-stripe this is a good idea please rework your logic based on a single flag. Looking forward to getting this into core! Let us know if you have any questions.

**Comment by tavish-stripe** [comment]
@hexfusion that seems like a reasonable solution. One alternative would be to provide the name and have `-ssl` get added in the TLS case. e.g. `-discovery-srv-name=foobar` would give `_foobar._tcp.service.consul` and `_foobar-ssl._tcp.service.consul`. This feels more flexible and readable to me, but either way solves the problem. What do you think? Happy to rework the patch either way.

**Comment by hexfusion** [comment]
@tavish-stripe I feel the definition should be explicit but to your point flexibility is a consideration.

defer to @xiang90

**Comment by hexfusion** [comment]
Another thing is that if we keep this `_etcd-server-ssl-${name}` this might make it less confusing as this is the previous standard and we are only appending name. Also a query of clusters might be easier to predefine/grep.

**Comment by tavish-stripe** [comment]
Yeah, that seems like a reasonable argument. I'll wait for @xiang90 's opinion and then rework the patch.

**Comment by hexfusion** [comment]
@tavish-stripe since @xiang90 has not added a comment I will assume he has no strong feelings. I would like to keep the existing format `etcd-server`, `etcd-server-ssl` +-{name}. Thank you and I look forward to our next review.

**Comment by xiang90** [comment]
sorry for the delay. i do not have strong feeling. just want to keep the flag down to one, and make it as clear as possible. we probably need to document it, and provide an example in the configuration doc as well.

**Comment by tavish-stripe** [comment]
@hexfusion @xiang90 I've reworked the patch to use the suffix. Please take a look and let me know if anything needs changing.



> we probably need to document it, and provide an example in the configuration doc as well.

Do I need to include that as part of this PR in the `Documentation/` folder, or is this something that should be handled as a separate PR?

**Comment by hexfusion** [comment]
> Do I need to include that as part of this PR in the Documentation/ folder, or is this something that should be handled as a separate PR?

@tavish-stripe please include in this PR as a single commit.

**Comment by xiang90** [comment]
@tavish-stripe 

probably the documentation should be added here https://github.com/coreos/etcd/blob/master/Documentation/op-guide/clustering.md#dns-discovery

**Comment by tavish-stripe** [comment]
@hexfusion @xiang90 I made a first cut of documentation for the new flag. Please let me know if there's anything I should change.

**Comment by xiang90** [comment]
@tavish-stripe This PR looks good to me in general. I would rely on @hexfusion and @gyuho or @spzala to give a final LGTM before merging.

Thanks.

File: Documentation/op-guide/configuration.md [hunk 3]
```diff
@@ -155,6 +155,11 @@ To start etcd automatically using custom settings at startup in Linux, using a [
 + default: ""
 + env variable: ETCD_DISCOVERY_SRV
 
+### --discovery-srv-name
++ Suffix to the DNS srv name queried when bootstrapping using DNS.
++ default: ""
++ env variable: ETCD_DISCOVERY_SRV
+
```
Comment (hexfusion) [inline Documentation/op-guide/configuration.md]: @tavish-stripe ETCD_DISCOVERY_SRV_NAME ?

**Comment by hexfusion** [comment]
@tavish-stripe one little knit, otherwise looks great!

**Comment by hexfusion** [comment]
:+1:

**Pull request merged**